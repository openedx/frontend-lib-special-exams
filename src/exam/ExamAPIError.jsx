import React from 'react';
import { useSelector } from 'react-redux';
import { getConfig } from '@edx/frontend-platform';
import { Alert, Hyperlink, Icon } from '@edx/paragon';
import { Info } from '@edx/paragon/icons';
import { injectIntl, intlShape, FormattedMessage } from '@edx/frontend-platform/i18n';
import messages from './messages';

const ExamAPIError = ({ intl }) => {
  const { SITE_NAME, SUPPORT_URL } = getConfig();
  const { apiErrorMsg } = useSelector(state => state.specialExams);
  const shouldShowApiErrorMsg = !!apiErrorMsg && !apiErrorMsg.includes('<');

  return (
    <Alert variant="danger" data-testid="exam-api-error-component">
      <Icon src={Info} className="alert-icon" />
      <Alert.Heading data-testid="error-details">
        {shouldShowApiErrorMsg ? apiErrorMsg : intl.formatMessage(messages.apiErrorDefault)}
      </Alert.Heading>
      <p>
        {SITE_NAME && SUPPORT_URL ? (
          <FormattedMessage
            id="exam.apiError.supportText.withLink"
            defaultMessage={
              'If the issue persists, please reach out to {supportLink} for assistance, '
              + 'and return to the exam once you receive further instructions.'
            }
            values={{
              supportLink: (
                <Hyperlink
                  data-testid="support-link"
                  destination={SUPPORT_URL}
                  target="_blank"
                >
                  {SITE_NAME} Support
                </Hyperlink>
              ),
            }}
          />
        ) : intl.formatMessage(messages.supportTextWithoutLink)}
      </p>
    </Alert>
  );
};

ExamAPIError.propTypes = {
  intl: intlShape.isRequired,
};

export default injectIntl(ExamAPIError);
