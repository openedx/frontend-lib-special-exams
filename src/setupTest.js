import 'babel-polyfill';
import '@testing-library/jest-dom';
import './data/__factories__';
import { getConfig, mergeConfig } from '@edx/frontend-platform';
import { configure as configureLogging } from '@edx/frontend-platform/logging';
import { configure as configureAuth, MockAuthService } from '@edx/frontend-platform/auth';
import { AppContext } from '@edx/frontend-platform/react';
import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import examReducer from './data/slice';

window.getComputedStyle = jest.fn(() => ({
  getPropertyValue: jest.fn(),
}));

class MockLoggingService {
  constructor() {
    this.logInfo = jest.fn();
    this.logError = jest.fn();
  }
}

export function initializeTestConfig() {
  mergeConfig({
    ...process.env,
    EXAMS_BASE_URL: process.env.EXAMS_BASE_URL || null,
  });
}

export function initializeMockApp() {
  initializeTestConfig();
  const loggingService = configureLogging(MockLoggingService, {
    config: getConfig(),
  });
  const authService = configureAuth(MockAuthService, {
    config: getConfig(),
    loggingService,
  });

  return { loggingService, authService };
}

let globalStore;

export function initializeTestStore(preloadedState = null, overrideStore = true) {
  let store = configureStore({
    reducer: {
      specialExams: examReducer,
    },
  });
  if (preloadedState) {
    store = configureStore({
      reducer: {
        specialExams: examReducer,
      },
      preloadedState,
    });
  }
  if (overrideStore) {
    globalStore = store;
  }
  return store;
}

function render(
  ui,
  {
    store = null,
    appContext = null,
    ...renderOptions
  } = {},
) {
  const Wrapper = ({ children }) => {
    const defaultAppContext = useMemo(() => ({
      authenticatedUser: {
        userId: 'abc123',
        username: 'Mock User',
        roles: [],
        administrator: false,
      },
    }), []);
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <AppContext.Provider value={appContext || defaultAppContext}>
        <IntlProvider locale="en">
          <Provider store={store || globalStore}>
            {children}
          </Provider>
        </IntlProvider>
      </AppContext.Provider>
    );
  };

  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// Re-export everything.
export * from '@testing-library/react';

// Override `render` method.
export {
  render,
};
