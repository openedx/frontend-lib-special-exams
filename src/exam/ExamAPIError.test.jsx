import '@testing-library/jest-dom';
import React from 'react';
import { getConfig } from '@edx/frontend-platform';
import { initializeTestStore, render } from '../setupTest';
import ExamAPIError from './ExamAPIError';

const originalConfig = jest.requireActual('@edx/frontend-platform').getConfig();
jest.mock('@edx/frontend-platform', () => ({
  ...jest.requireActual('@edx/frontend-platform'),
  getConfig: jest.fn(),
}));
getConfig.mockImplementation(() => originalConfig);

describe('ExamAPIError', () => {
  const defaultMessage = 'A system error has occurred with your exam.';

  let store;

  beforeEach(() => {
    store = initializeTestStore();
  });

  it('renders with the default information', () => {
    store.getState = () => ({ specialExams: {} });

    const tree = render(
      <ExamAPIError />,
      { store },
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders support link if site name and support url are given', () => {
    const config = {
      SITE_NAME: 'Open edX',
      SUPPORT_URL: 'https://support.example.org/',
    };
    getConfig.mockImplementation(() => config);

    store.getState = () => ({ specialExams: {} });

    const { getByTestId } = render(
      <ExamAPIError />,
      { store },
    );

    const supportLink = getByTestId('support-link');
    expect(supportLink).toHaveProperty('href', config.SUPPORT_URL);
    expect(supportLink).toHaveTextContent(`${config.SITE_NAME} Support`);
  });

  it('renders error details when provided', () => {
    store.getState = () => ({
      specialExams: { apiErrorMsg: 'Something bad has happened' },
    });

    const { queryByTestId } = render(
      <ExamAPIError />,
      { store },
    );

    expect(queryByTestId('error-details')).toHaveTextContent(store.getState().specialExams.apiErrorMsg);
  });

  it('renders default message when error is HTML', () => {
    store.getState = () => ({
      specialExams: { apiErrorMsg: '<Response is HTML>' },
    });

    const { queryByTestId } = render(
      <ExamAPIError />,
      { store },
    );

    expect(queryByTestId('error-details')).toHaveTextContent(defaultMessage);
  });

  it('renders default message when there is no error message', () => {
    store.getState = () => ({
      specialExams: { apiErrorMsg: '' },
    });

    const { queryByTestId } = render(
      <ExamAPIError />,
      { store },
    );

    expect(queryByTestId('error-details')).toHaveTextContent(defaultMessage);
  });
});
