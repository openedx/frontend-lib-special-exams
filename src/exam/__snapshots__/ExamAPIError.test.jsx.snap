// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExamAPIError renders with the default information 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fade alert-content alert alert-danger show"
        data-testid="exam-api-error-component"
        role="alert"
      >
        <div
          class="pgn__alert-message-wrapper"
        >
          <div
            class="alert-message-content"
          >
            <span
              class="pgn__icon alert-icon"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <div
              class="alert-heading h4"
              data-testid="error-details"
            >
              A system error has occurred with your exam.
            </div>
            <p>
              If the issue persists, please reach out to support for assistance, and return to the exam once you receive further instructions.
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fade alert-content alert alert-danger show"
      data-testid="exam-api-error-component"
      role="alert"
    >
      <div
        class="pgn__alert-message-wrapper"
      >
        <div
          class="alert-message-content"
        >
          <span
            class="pgn__icon alert-icon"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="alert-heading h4"
            data-testid="error-details"
          >
            A system error has occurred with your exam.
          </div>
          <p>
            If the issue persists, please reach out to support for assistance, and return to the exam once you receive further instructions.
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
